src/source/Signaling/LwsApiCalls.c

```diff
@@ -294,8 +294,15 @@ INT32 lwsWssCallbackRoutine(struct lws* wsi, enum lws_callback_reasons reason, P
             ATOMIC_STORE(&pSignalingClient->result, (SIZE_T) SERVICE_CALL_UNKNOWN);
 
             if (connected && !ATOMIC_LOAD_BOOL(&pSignalingClient->shutdown)) {
-                // Handle re-connection in a reconnect handler thread
-                CHK_STATUS(THREAD_CREATE(&pSignalingClient->reconnecterTracker.threadId, reconnectHandler, (PVOID) pSignalingClient));
+                // Handle re-connection in a reconnect handler thread. Set the terminated indicator before the thread
+                // creation and the thread itself will reset it. NOTE: Need to check for a failure and reset.
+                ATOMIC_STORE_BOOL(&pSignalingClient->reconnecterTracker.terminated, FALSE);
```

Comment:
this would definitely improve, but I am still concerned that the race could happen with little chance, as we cannot batch loading `shutdown` and clearing `terminated`in an atomic operation, so for magic reason, we loaded the `shutdown` flag, but have not cleared `terminated` yet, then freeSignaling starts and finishes.

Reply:
They can't be indeed be processed atomically. The main difficulty and the complexity of the code comes from the fact that there is no platform independent way to await on the thread handle like in Windows so we have to do this insane interlocking/atomics operations to guarantee ordering.

In this case, there is a chance indeed that the shutdown could be set but the terminated for the reconnect tracker is not. However. there will be the listener tracker thats currently running the event loop which would then block. Let me check if there is a chance for that to misbehave.

Regardless of that, these changes should go in and we should in parallel track if there is still an issue with the shutdown sequence on edge cases like reconnect

Reply:
Yeah, the freeSignaling attempts to tear down the listener first (which is driving the reconnect scenario that spawns the reconnect handler).

Reply:
In that case, how about if we remove "terminated" flag in the thread tracker, and have an atomic integer in the signaling client that contains the three states: signaling shutdown, reconnect thread terminated, and listener thread terminated? Since these are just booleans, we only need 3 bits to represent all states. 

This way, we'll be able to atomically load and store the three states in batch.

Reply:
what would happen to the ongoing reconnect handler when the listener tracker is teared down?
sorry for too many questions, just like to learn more about the multi-threading in the SDK.

Reply:
The atomicity is not really about two reads. It's more about multi-read-check-write. In either case, there is a chance for the race if you just consider this local example. 

A better solution perhaps is a semaphore similar to what we have in PIC/C Producer transitions. In this case, every time there is a source of liveness, we increment a semaphore (listener, ICE refresher, reconnecter). The shutdown would then flag and await the semaphore with a timeout (instead of the awaitForThreadTermination API sequence).

The semaphore would not remove the need for the existing primitives like the tracker as we need to do certain synchronization for listener/reconnect threading.

Reply:
oh, you are saying the listenerTracker would be blocked if the reconnect handler is about to start?

Reply:
> what would happen to the ongoing reconnect handler when the listener tracker is teared down?
> sorry for too many questions, just like to learn more about the multi-threading in the SDK.

The reconnect is getting kicked off from the event loop. The event loop is powered by the listener thread. The reconnect gets kicked off on error and on disconnect callback handling to ensure we have a way to reconnect. We can't do the reconnection from within the existing event loop/websocket so we need a new source of liveness.

The reconnect thread would then launch and block until the listener thread frees the lock on exit, meaning that the reconnect thread could then go and drive the state machinery and exit. Part of the state machinery state is to launch a new listener thread.

So, the net result is old listener -> reconnect if an error -> reconnect handler wait for old listener to exit and free the lock -> old listener exits and unlocks the lock -> reconnect thread drives the state machinery -> kicks off a new listener -> reconnect thread exits.

Reply:
> The atomicity is not really about two reads. It's more about multi-read-check-write. In either case, there is a chance for the race if you just consider this local example.

Right, I was thinking to use CAS for this, which allows us to do multi-read-check-write.

> A better solution perhaps is a semaphore similar to what we have in PIC/C Producer transitions. In this case, every time there is a source of liveness, we increment a semaphore (listener, ICE refresher, reconnecter). The shutdown would then flag and await the semaphore with a timeout (instead of the awaitForThreadTermination API sequence).

I was thinking about using a semaphore as well, but wouldn't it face the same problem as there's no way to update the shutdown flag and increment the semaphore at the same time?

Reply:
> oh, you are saying the listenerTracker would be blocked if the reconnect handler is about to start?

The listener thread is the one that is powering the libWebSocket threading and the event loop. The callback is primarily (though, not always) is happening from the thread that is actually powering the event loop in lwsCompleteSync

Reply:
> > The atomicity is not really about two reads. It's more about multi-read-check-write. In either case, there is a chance for the race if you just consider this local example.
> 
> Right, I was thinking to use CAS for this, which allows us to do multi-read-check-write.
> 
> > A better solution perhaps is a semaphore similar to what we have in PIC/C Producer transitions. In this case, every time there is a source of liveness, we increment a semaphore (listener, ICE refresher, reconnecter). The shutdown would then flag and await the semaphore with a timeout (instead of the awaitForThreadTermination API sequence).
> 
> I was thinking about using a semaphore as well, but wouldn't it face the same problem as there's no way to update the shutdown flag and increment the semaphore at the same time?

There are no atomic primitives for this type of CAS. You will need interlocking.

The blocking rw semaphore we have is the right primitive. Readers would increment the lock count, the writer (shutdown sequence) would block until the semaphore value is 0 and continue. All this time no new readers will be permitted. 

Still, there is chance for a timeout (due to the complexity in some edge cases when the socket dies) and we might end up with timing out to continue the tear down which in some cases could lead to a crash. This is still better than a hang

Reply:
I would still prefer to go with these changes and cut an enhancement work item for the semaphore solution

Reply:
alright, good to learn all the details, i think we all on the same page regarding the race we need to fix.
what would be the perfect solution? is this PR good enough? if so, i would encourage to get it merged asap, then i could try it out too.

Reply:
This PR should fix the issue with the teardown race for the issue we had identified earlier and the need to increase the shutdown timeout (most important part of the fix actually).

As mentioned, there is a theoretical chance of a race but that particular issue is handled at the higher level logic by which the shutdown tears down the listener first and by that time the reconnect would have a chance to react on the shutdown and exit. In this case, there is no case with the race as the shutdown thread would go and block on the termination flag for the reconnect.

However, this is still brittle and handles just this case and perhaps we have other cases there. I would prefer a better solution for the shutdown sequence based not on a shutdown flag (atomic) but rather on a semaphore. That should be a separate work item though